adr/ADR-53.md

```diff
@@ -0,0 +1,147 @@
+
+Specifically, we have no way to guarantee a write like `kv.Put` can be observed by a subsequent `kv.Get` or `kv.Watch`,
+especially when the KV/stream is replicated or mirrored.
+
+## Context
+
+The topic of immediate consistency within NATS JetStream can sometimes be a bit confusing. On our docs we claim we
+maintain immediate consistency (as opposed to eventual consistency) even in the face of failures. Which is true.. but,
+as with anything, it depends.
+
+- **Monotonic writes**, all writes to a single stream (replicated or not) are monotonic. It's ordered regardless of
+  publisher by the stream sequence.
+- **Monotonic reads**, if you're using consumers. All reads for a consumer (replicated or not) are monotonic. It's
+  ordered by consumer delivery sequence. (Messages can be redelivered on failure, but this also depends on which
+  settings are used)
+
+Those paths are immediately consistent.. but they are not immediately consistent with respect to each other. This is no
+problem for publishers and consumers of a stream, because they observe all operations to be monotonic.
+But, if you use the KV abstraction for example, you're more often going to use single message gets through `kv.Get`.
+Since those rely on `DirectGet`, even followers can answer, which means we (by default) can't guarantee read-after-write
+or even monotonic reads. Such message GET requests get served randomly by all servers within the peer group (or even
+mirrors if enabled). Those obviously can't be made immediately consistent, since both replication and mirroring are
+async.
+
+Also, when following up a `kv.Create` with `kv.Keys`, you might expect read-after-write such that the returned keys
+contains the key you've just written to. This also requires read-after-write.
+
+## Design
+
+Before sharing the proposed design, let's look at an alternative. Read-after-write could be achieved by having reads (on
+an opt-in basis) go through Raft replication first. This has several disadvantages:
+
+- Reads will become significantly slower, due to requiring replication first.
+- Reads require quorum, due to replication, disallowing any reads when there's downtime or temporarily no leader.
+- Only the stream leader can answer reads, as it is the first one to know that it can answer the request. (Followers
+  replicate asynchronously, so letting them answer would make the response take even longer to return.)
+- Mirrors can still answer `DirectGet` requests, the transparency of mirrors answering read requests will violate any
+  read-after-write guarantees (as the client will not know). This would mean mirrors must not be enabled if this
+  guarantee should be kept.
+- Read-after-write guarantees could temporarily be violated when scaling streams up or down.
+- This is not a compatible approach for consumers, meaning they could not have these guarantees based on this approach.
```

Comment:
would be good to clarify this one

Reply:
Have added:
>  It would require limiting consumer creation to R1 on the stream leader, which is not possible since the assignment is  done by the meta leader that has no knowledge about the stream leader. A replicated consumer could violate the requirement if the consumer leader changes to an outdated follower in between.